src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -271,71 +304,194 @@ private async Task WriteDataAsyncAwaited(int streamId, StreamOutputFlowControl f
             await ThreadPoolAwaitable.Instance;
         }
 
+        /* https://tools.ietf.org/html/rfc7540#section-6.9
+            +-+-------------------------------------------------------------+
+            |R|              Window Size Increment (31)                     |
+            +-+-------------------------------------------------------------+
+        */
         public Task WriteWindowUpdateAsync(int streamId, int sizeIncrement)
         {
             lock (_writeLock)
             {
+                if (_completed)
+                {
+                    return Task.CompletedTask;
+                }
+
                 _outgoingFrame.PrepareWindowUpdate(streamId, sizeIncrement);
-                return WriteFrameUnsynchronizedAsync();
+                WriteHeaderUnsynchronized();
+                var buffer = _outputWriter.GetSpan(4);
+                Bitshifter.WriteUInt31BigEndian(buffer, (uint)sizeIncrement);
+                _outputWriter.Advance(4);
+                return _flusher.FlushAsync();
             }
         }
 
+        /* https://tools.ietf.org/html/rfc7540#section-6.4
+            +---------------------------------------------------------------+
+            |                        Error Code (32)                        |
+            +---------------------------------------------------------------+
+        */
         public Task WriteRstStreamAsync(int streamId, Http2ErrorCode errorCode)
         {
             lock (_writeLock)
             {
+                if (_completed)
+                {
+                    return Task.CompletedTask;
+                }
+
                 _outgoingFrame.PrepareRstStream(streamId, errorCode);
-                return WriteFrameUnsynchronizedAsync();
+                WriteHeaderUnsynchronized();
+                var buffer = _outputWriter.GetSpan(4);
```

Comment:
These three lines writing a uint/31-bit int/short could be extracted to extension methods.

Reply:
If there were duplicates it would be worth extracting, but 
every usage of WriteUInt32BigEndian ends up being a little different.